### PR TITLE
Towards retrieval of maximum keepbits

### LINKED ISSUE #22: Retrieval of maximum keepbits
The bitwise information content can depend on the dimension `bitinformation` is applied to. Changes along dimensions like `lon` and `time` are rather small compared to those along the `lat` dimension. This might depend on the dataset though and the user might want to run the analysis across each dimension individually to identify the dimension with most keepbits and use those.

I propose to allow `get_bitinformation` handle a list of dims or a string like `maxdim` to return the keepbits of the dimension with the largest information content.

### PR DESCRIPTION
**Pull request by observingClouds**
Eventually closes #22
To do so, it adds changes to the info_per_bits dataset and continues #79

### CODE REVIEW

**Comment by observingClouds** [comment]
This is still a draft! I haven't tested it thoroughly yet.

**Comment by observingClouds** [comment]
Especially edge-cases, e.g. when not all variables in a dataset have the same amount of dimensions needs some tests.

File: tests/test_bitinformation_pipeline.py [hunk 1]
```diff
@@ -27,6 +27,18 @@ def test_full():
...
 
 
+def test_full_max_keepbits():
+    """Test pipeline to get maximum keepbits"""
+    label = "air_temperature"
+    ds = xr.tutorial.load_dataset(label)
+    bi = xb.get_bitinformation(ds)
+    kb = xb.get_keepbits(bi)
+    kb_max = kb.max(dim="dim")
+    _ = xb.plot_bitinformation(bi.isel(dim=slice(0, 1)))
```
Comment (aaronspring) [inline tests/test_bitinformation_pipeline.py:37]: why slice(0,1)?

File: tests/test_get_bitinformation.py [hunk 2]
```diff
@@ -132,4 +132,29 @@ def test_get_bitinformation_dtype(rasm, dtype):
...
+        dtype_bits
+    )
+
+
+def test_get_bitinformation_multidim(rasm):
+    """Test xb.get_bitinformation runs on all dimensions by default"""
+    ds = rasm
+    bi = xb.get_bitinformation(ds)
+    # check length of dimension
+    assert bi.dims["dim"] == 3
```
Comment (aaronspring) [inline tests/test_get_bitinformation.py:146]:
```suggestion
    assert bi.dims["dim"] == len(ds.dims)
```

File: tests/test_get_bitinformation.py [hunk 3]
```diff
@@ -132,4 +132,29 @@ def test_get_bitinformation_dtype(rasm, dtype):
...
+    assert any(bi_time != bi_y)
+    assert any(bi_y != bi_x)
+
+
+def test_get_bitinformation_different_variables_dims(rasm):
+    """Test xb.get_bitinformation runs with variables of different dimensionality"""
+    ds = rasm
+    # add variable with different dimensionality
+    ds["Tair_mean"] = ds.Tair.mean(dim="time")
+    xb.get_bitinformation(ds)
```
Comment (aaronspring) [inline tests/test_get_bitinformation.py:160]: could you add the `assert` statement with what to expect? `Tair_mean` dropped?

File: xbitinfo/bitround.py [hunk 4]
```diff
@@ -37,18 +37,25 @@ def _keepbits_interface(da, keepbits):
         else:
             raise ValueError(f"name {v} not for in keepbits: {keepbits.keys()}")
     elif isinstance(keepbits, xr.Dataset):
+        assert keepbits.coords["inflevel"].shape <= (
+            1,
+        ), "length of 'threshold' dimension of keepbits need to be one"
```
Comment (aaronspring) [inline xbitinfo/bitround.py:42]: zero or one? is a non-dimensional coord also allowed? I think so

File: xbitinfo/xbitinfo.py [hunk 5]
```diff
@@ -123,31 +126,37 @@ def get_bitinformation(ds, dim=None, axis=None, label=None, overwrite=False, **k
         >>> ds = xr.tutorial.load_dataset("air_temperature")
         >>> xb.get_bitinformation(ds, dim="lon")
         <xarray.Dataset>
-        Dimensions:  (bit32: 32)
+        Dimensions:  (bit32: 32, dim: 1)
```
Comment (aaronspring) [inline xbitinfo/xbitinfo.py:129]: do we need to have this as an extra dimension if only one dim? could also squeeze. we need the additional dimension fo more than one dim, but for one dim only I'd prefer to squeeze the dim to a coord

File: xbitinfo/xbitinfo.py [hunk 6]
```diff
@@ -123,31 +126,37 @@ def get_bitinformation(ds, dim=None, axis=None, label=None, overwrite=False, **k
...
+            except FileNotFoundError:
+                logging.info(
+                    f"No bitinformation could be found for {label}. Recalculating..."
+                )
+
         # check keywords
-        if (axis is None and dim is None) or (axis is not None and dim is not None):
+        if axis is not None and dim is not None:
             raise ValueError(
                 "Please provide either `axis` or `dim` but not both or none."
```
Comment (aaronspring) [inline xbitinfo/xbitinfo.py:161]:
```suggestion
                "Please provide either `axis` or `dim` but not both."
```

**Review by aaronspring — APPROVED** [review]
really nice. the only comment to discuss concerns whether `dim` should be squeezed from dimension to coordinate from `info_per_bit` if only one element?

File: tests/test_bitinformation_pipeline.py (hunk 1 above)
Comment (observingClouds) [reply tests/test_bitinformation_pipeline.py:37]: To keep the `dim`-dimension. I will add also the case, when `dim` only remains as a coordinate.

File: xbitinfo/bitround.py (hunk 4 above)
Comment (observingClouds) [reply xbitinfo/bitround.py:42]: Yes, both are consciously allowed.

File: tests/test_bitinformation_pipeline.py (hunk 1 above)
Comment (aaronspring) [reply tests/test_bitinformation_pipeline.py:37]: ah ok. I usually do `.isel(dim=[0])` which is shorter. I think it would be easier if dim doesnt need to be a dimension here.

File: xbitinfo/bitround.py (hunk 4 above)
Comment (aaronspring) [reply xbitinfo/bitround.py:42]: I would also rename threshold to inflevel.

Comment (aaronspring) [reply xbitinfo/bitround.py:42]: `dimension ... need to be one` sounds like it needs to be a dimension but non-dim coord also works

File: tests/test_bitinformation_pipeline.py (hunk 1 above)
Comment (observingClouds) [reply tests/test_bitinformation_pipeline.py:37]: `dim` does not need to be a dimension here. I just added it to make sure to test that both cases work. (though I still need to add the other test case as well)

**Comment by observingClouds** [comment]
The benefit of keeping `dim` a dimension even in the case this can be squeezed is the ability to easily merge datasets with different amount of dimensions.

```python
import xbitinfo as xb
import xarray as XR

ds = xr.tutorial.load_dataset("rasm")
ds['Tair_mean'] = ds.Tair.mean(dim='time')

info_per_bit_time = xb.get_bitinformation(ds, dim='time')
info_per_bit_x = xb.get_bitinformation(ds, dim='x')

# works
print(xr.merge([info_per_bit_time, info_per_bit_x]))
"""
<xarray.Dataset>
Dimensions:    (dim: 2, bit64: 64)
Coordinates:
  * dim        (dim) <U4 'time' 'x'
  * bit64      (bit64) <U3 '±' 'e1' 'e2' 'e3' 'e4' ... 'm49' 'm50' 'm51' 'm52'
Data variables:
    Tair       (dim, bit64) float64 0.431 0.7688 0.0 ... 0.07677 0.08063 0.08087
    Tair_mean  (dim, bit64) float64 nan nan nan nan ... 0.09126 0.04861 0.05123
Attributes:
    xbitinfo_description:  bitinformation calculated by xbitinfo.get_bitinfor...
    python_repository:     https://github.com/observingClouds/xbitinfo
    julia_repository:      https://github.com/milankl/BitInformation.jl
    reference_paper:       http://www.nature.com/articles/s43588-021-00156-2
    xbitinfo_version:      0.0.0
"""

# doesn't work
print(xr.merge([info_per_bit_time.squeeze(), info_per_bit_x.squeeze()]))
"""
MergeError: conflicting values for variable 'dim' on objects to be combined. You can skip this check by specifying compat='override'.
"""

# alternate approach drops Tair_mean
print(xr.concat([info_per_bit_time.squeeze(), info_per_bit_x.squeeze()], dim='dim'))
"""
<xarray.Dataset>
Dimensions:  (bit64: 64, dim: 2)
Coordinates:
  * bit64    (bit64) <U3 '±' 'e1' 'e2' 'e3' 'e4' ... 'm49' 'm50' 'm51' 'm52'
  * dim      (dim) <U4 'time' 'x'
Data variables:
    Tair     (dim, bit64) float64 0.431 0.7688 0.0 ... 0.07677 0.08063 0.08087
Attributes:
    xbitinfo_description:  bitinformation calculated by xbitinfo.get_bitinfor...
    python_repository:     https://github.com/observingClouds/xbitinfo
    julia_repository:      https://github.com/milankl/BitInformation.jl
    reference_paper:       http://www.nature.com/articles/s43588-021-00156-2
    xbitinfo_version:      0.0.0
"""
```

But maybe it is better to align with the xarray behaviour, where e.g. `ds['Tair_mean']` has `time` only as a coordinate and the dimension is dropped.

**Comment by aaronspring** [comment]
> But maybe it is better to align with the xarray behaviour, where e.g. ds['Tair_mean'] has time only as a coordinate and the dimension is dropped.

I agree.
I usually take the `xr.concat` path.
